src/foam/u2/property/MDPopup.js

```diff
@@ -0,0 +1,241 @@
+/**
+* @license
+* Copyright 2020 The FOAM Authors. All Rights Reserved.
+* http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+
+foam.CLASS({
+  package: 'foam.u2.property',
+  name: 'MDPopup',
+  extends: 'foam.u2.View',
+
+  requires: [
+    'foam.u2.property.MenuElement'
+  ],
+
+  imports: [
+    'clearTimeout',
+    'window',
+    'setTimeout',
+  ],
+
+  exports: [
+    'as popup'
+  ],
+
+  documentation: 'A floating menu that pops up, positioning itself over the ' +
+      'element that launches it with $$DOC{ref:".open"}.',
+
+  properties: [
+    ['hMargin', 8],
+    ['vMargin', 8],
+    ['maxDisplayCount', 5],
+    ['itemHeight', 48],
+    ['itemWidth', 100],
+    ['isHidden', true],
+    ['removeTimeout', 0],
+    ['isClosing', false],
+    'choices',
+    'data',
+    'index'
+  ],
+
+  methods: [
+    function open(index, sourceElement) {
+      /* Launches the menu, with the given selected item index, animating out
+         from the given sourceElement's client rect. */
+      var startingClientRect = sourceElement.getBoundingClientRect();
+      var vp = {
+        height: this.window.innerHeight || this.document.documentElement.clientHeight,
+        width: this.window.innerWidth || this.document.documentElement.clientWidth
+      };
+      this.itemHeight = startingClientRect.height;
+      this.itemWidth = startingClientRect.width - 16;
+
+      var pxAbove = startingClientRect.top - this.vMargin - 4;
+      var pxBelow = vp.height - startingClientRect.bottom - this.vMargin - 4;
+
+      // "Slots" represent potential screen real estate for drawing the menu.
+      var slotsAbove = Math.floor((pxAbove > 0) ? pxAbove / this.itemHeight : 0);
+      var slotsBelow = Math.floor((pxBelow > 0) ? pxBelow / this.itemHeight : 0);
+      // "Items" are the menu items going into these slots.
+      var itemsAbove = index;
+      var itemsBelow = this.choices.length - index - 1;
+
+      // Show as many choices as there is room for, capped by how many we have
+      // and by the maxDisplayCount (usually 5).
+      var menuCount = Math.min(this.choices.length, this.maxDisplayCount,
+          slotsAbove + slotsBelow + 1);
+      var halfMenuCount = Math.floor(menuCount / 2);
+
+      // If scrolling, this becomes the scroll offset.
+      var itemForFirstSlot = 0;
+      // If the selecteditem can't be in the best place, we animate it from the
+      // start rect by this many slots. Negative offset means move up.
+      var selectedOffset = 0;
+
+      if (menuCount < this.choices.length) { // Scrolling required.
```

Comment:
Fix spacing, here and other if's in file.
if ( menuCount < this.choices.length ) {